File: day8/solution.zig
const std = @import("std");
const input = @embedFile("input.txt");

const Position = @Vector(3, i64);
const Pair = struct { pos1_idx: usize, pos2_idx: usize, dst: u64 };

fn cmp_pairs(_: void, a: Pair, b: Pair) bool {
    return a.dst < b.dst;
}

pub fn find(parents: []usize, idx: usize) usize {
    if (idx == parents[idx]) {
        return idx;
    }
    parents[idx] = find(parents, parents[idx]);
    return parents[idx];
}

pub fn merge(parents: []usize, idx1: usize, idx2: usize) void {
    parents[find(parents, parents[idx1])] = find(parents, idx2);
}

pub fn main() void {
    const allocator = std.heap.c_allocator;
    const line_count = std.mem.countScalar(u8, input, '\n');
    const numbers: []Position = allocator.alloc(Position, line_count) catch unreachable;
    defer allocator.free(numbers);

    var timer = std.time.Timer.start() catch unreachable;

    var pairs = std.ArrayList(Pair).initCapacity(allocator, line_count * line_count) catch unreachable;
    defer pairs.deinit(allocator);

    const parents: []usize = allocator.alloc(usize, line_count) catch unreachable;
    defer allocator.free(parents);

    var row_iter = std.mem.splitScalar(u8, input, '\n');

    var idx: usize = 0;
    while (row_iter.next()) |row| {
        if (row.len == 0) continue;

        var coord: [3]i64 = undefined;

        var numbers_iter = std.mem.splitScalar(u8, row, ',');
        for (0..3) |i| {
            const num_str = numbers_iter.next() orelse unreachable;
            const num = std.fmt.parseInt(i64, num_str, 10) catch unreachable;
            coord[i] = num;
        }

        numbers[idx] = @bitCast(coord);
        parents[idx] = idx;
        idx += 1;
    }

    for (numbers, 0..) |pos, i| {
        for (numbers, 0..) |pos2, j| {
            if (i > j) {
                pairs.appendAssumeCapacity(.{ .pos1_idx = i, .pos2_idx = j, .dst = dst(pos, pos2) });
            }
        }
    }

    std.mem.sort(Pair, pairs.items, {}, cmp_pairs);

    var last_merged: struct { idx1: usize, idx2: usize } = undefined;
    for (pairs.items, 0..) |pair, i| {
        if (i + 1 == 1000) {
            const counts: []usize = allocator.alloc(usize, line_count) catch unreachable;
            defer allocator.free(counts);
            @memset(counts, 0);

            for (0..numbers.len) |j| {
                counts[find(parents, j)] += 1;
            }

            std.mem.sort(usize, counts, {}, std.sort.desc(usize));

            const elapsed_part_one = timer.read();
            std.debug.print("Part one: {d} in {d}µs\n", .{ counts[0] * counts[1] * counts[2], elapsed_part_one / 1000 });
        }
        if (find(parents, pair.pos1_idx) != find(parents, pair.pos2_idx)) {
            merge(parents, pair.pos1_idx, pair.pos2_idx);
            last_merged = .{ .idx1 = pair.pos1_idx, .idx2 = pair.pos2_idx };
        }
    }
    const elapsed_part_two = timer.read();
    std.debug.print("Part two: {d} in {d}µs\n", .{ numbers[last_merged.idx1][0] * numbers[last_merged.idx2][0], elapsed_part_two / 1000 });
}

fn dst(a: Position, b: Position) u64 {
    return @abs(@reduce(.Add, (b - a) * (b - a)));
}

File: day8/solution.ml
type position = { x : int; y : int; z : int }

let positions =
  Arg.read_arg "input.txt" |> Array.to_list
  |> List.map (fun line ->
      match String.split_on_char ',' line with
      | [ x; y; z ] ->
          { x = int_of_string x; y = int_of_string y; z = int_of_string z }
      | _ -> failwith "unexpected input")

let pos_count = List.length positions

let dst p1 p2 =
  let dx = p2.x - p1.x in
  let dy = p2.y - p1.y in
  let dz = p2.z - p1.z in
  (dx * dx) + (dy * dy) + (dz * dz)

let unique_pairs positions =
  let rec unique_pairs_i i = function
    | [] -> []
    | x :: xs ->
        let pairs = List.mapi (fun j y -> (i, j + i + 1, dst x y)) xs in
        pairs @ unique_pairs_i (i + 1) xs
  in
  unique_pairs_i 0 positions

let rec find parents idx =
  if idx = parents.(idx) then idx
  else (
    parents.(idx) <- find parents parents.(idx);
    parents.(idx))

let merge parents a b = parents.(find parents parents.(a)) <- find parents b

let sorted =
  List.sort
    (fun (_, _, dst_a) (_, _, dst_b) -> compare dst_a dst_b)
    (unique_pairs positions)

let parents = Array.init pos_count (fun i -> i)

let part_two =
  Array.of_list sorted
  |> Array.fold_left
       (fun (last_merged, i) (pos_a_i, pos_b_i, _) ->
         if i + 1 = 1000 then (
           let counts = Array.make pos_count 0 in
           Array.iteri
             (fun j _ ->
               let idx = find parents j in
               counts.(idx) <- counts.(idx) + 1)
             counts;
           Array.sort (fun x y -> compare y x) counts;
           Printf.printf "Part one: %d\n" (counts.(0) * counts.(1) * counts.(2)));

         if find parents pos_a_i <> find parents pos_b_i then (
           merge parents pos_a_i pos_b_i;
           ((pos_a_i, pos_b_i), i + 1))
         else (last_merged, i + 1))
       ((0, 0), 0)
  |> fun ((a, b), _) ->
  let pos_arr = Array.of_list positions in
  Printf.printf "Part two: %d * %d = %d\n" pos_arr.(a).x pos_arr.(b).x
    (pos_arr.(a).x * pos_arr.(b).x)
